library/src/tests/resources/state_preparation.qs

```diff
@@ -0,0 +1,137 @@
+        // Ucompute minus
+        H(q);
+        X(q);
+        if ( not CheckZero(q) ) {
+            fail "|-> preparation failed.";
+        }
+    }
+
+    operation TestBellState(): Unit {
+        use q = Qubit[2];
+        PreparePureStateD([Sqrt(0.5), 0.0, 0.0, Sqrt(0.5)], q);
+        // Ucompute Bell
+        CNOT(q[0], q[1]);
+        H(q[0]);
+        if ( not CheckAllZero(q) ) {
+            fail "Bell state preparation failed.";
+        }
+    }
+
+    operation TestCat3State(): Unit {
+        use q = Qubit[3];
+        PreparePureStateD([Sqrt(0.5), 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, Sqrt(0.5)], q);
+        // Ucompute Cat
+        CNOT(q[0], q[2]);
+        CNOT(q[0], q[1]);
+        H(q[0]);
+        if ( not CheckAllZero(q) ) {
+            fail "Cat state preparation failed.";
+        }
+    }
+
+    operation PrepareComplex(qs : Qubit[]) : Unit is Adj {
+        H(qs[0]);
+        S(qs[0]);
+        H(qs[1]);
+        T(qs[1]);
+    }
+
+    operation TestPrepareComplex(): Unit {
+        use q = Qubit[2];
+        let c00 = ComplexPolar(0.5, 0.0);
+        let c01 = ComplexPolar(0.5, PI()/4.0);
+        let c10 = ComplexPolar(0.5, PI()/2.0);
+        let c11 = ComplexPolar(0.5, 3.0*PI()/4.0);
+        ApproximatelyPreparePureStateCP(0.0, [c00, c01, c10, c11], q);
+        Adjoint PrepareComplex(q);
+        if ( not CheckAllZero(q) ) {
+            fail "Complex state preparation failed.";
+        }
+    }
+
+    operation TestPreparationCompletion(): Unit {
+        let testCases = [
+            // Test positive coefficients
+            [0.773761, 0.633478],
+            [0.183017, 0.406973, 0.604925, 0.659502],
+            [0.0986553, 0.359005, 0.465689, 0.467395, 0.419893, 0.118445, 0.461883, 0.149609],
+            [0.271471, 0.0583654, 0.11639, 0.36112, 0.307383, 0.193371, 0.274151, 0.332542, 0.130172, 0.222546, 0.314879, 0.210704, 0.212429, 0.245518, 0.30666, 0.22773],
+
+            // Test negative coefficients; should give same probabilities as positive coefficients
+            [-0.773761, 0.633478],
+            [0.183017, -0.406973, 0.604925, 0.659502],
+            [0.0986553, -0.359005, 0.465689, -0.467395, 0.419893, 0.118445, -0.461883, 0.149609],
+            [-0.271471, 0.0583654, 0.11639, 0.36112, -0.307383, 0.193371, -0.274151, 0.332542, 0.130172, 0.222546, 0.314879, -0.210704, 0.212429, 0.245518, -0.30666, -0.22773],
+
+            // Test unnormalized coefficients
+            [1.0986553, 0.359005, 0.465689, -0.467395, 0.419893, 0.118445, 0.461883, 0.149609],
+
+            // Test missing coefficients
+            [1.0986553, 0.359005, 0.465689, -0.467395, 0.419893, 0.118445]
+        ];
+
+        for coefficients in testCases {
+            let L = Length(coefficients);
+            let N = Ceiling(Log(IntAsDouble(L))/LogOf2() - 0.001);
+            use q = Qubit[N];
+            PreparePureStateD(coefficients, q);
+            // NOTE: We cannot check that the state is actually correct.
```

Comment:
We actually can check the state from Rust using DumpMachine and the registered receiver. We can make a new test utility to help with this. Let’s sync up on that one, I think it will be handy going forward, and might even be something we can adopt for the intrinsic tests.

Reply:
Yes, it's about time!

Reply:
I tried out a couple alternatives for how to verify dump output and finally found one I like best. You can check out the change here: https://github.com/microsoft/qsharp/commit/5a31cc7b5d46adddcd886a875a930efee1250b33. Feel free to cherry-pick that commit if you want to get it into your branch or we can discuss more if you want to go over it before adopting the changes.

Reply:
I think it's a great utility to be able to verify the quantum state in tests cases. I suggest to consider numerical robustness. Some internal changes to the operation or to other operations that it's referring to could add some negligible numerical imprecision that changes the output. Also it could add a global phase, which doesn't change the quantum state, but the way it's going to be displayed. Further, there may be test cases in which one wants to compare a subset of amplitudes.

Reply:
We are inspecting the state now as displayed by DumpMachine. Its precision is only about 10^-4, but this should be enough for the test cases we currently have.